### PR TITLE
feat(manager-location): add endpoints to assign and unassign managers to locations

### PR DESCRIPTION
**Pull request by eugeneanokye99**
## What
Implements Epic 3 location/department management and related auth/security hardening needed for operability.

### Delivered in this PR
- **US-LOC-01 / FR-LOC-01**
  - `POST /api/v1/locations` (HR_ADMIN)
  - `GET /api/v1/locations` (active locations)
  - `PATCH /api/v1/locations/{id}` (HR_ADMIN)
  - Duplicate location name handling (`409`)
- **US-LOC-02 / FR-LOC-02**
  - `POST /api/v1/locations/{locationId}/departments` (HR_ADMIN)
  - `GET /api/v1/locations/{locationId}/departments` (HR_ADMIN, MANAGER)
  - Department uniqueness scoped to location (`409`)
  - `404` for invalid `locationId`
- **US-LOC-03 / FR-LOC-03**
  - `GET /api/v1/managers/me/locations` (MANAGER)
  - Returns empty list when no assignments
- **Operational extension to close assignment gap in FR-LOC-03**
  - `POST /api/v1/managers/{managerEmployeeId}/locations/{locationId}` (HR_ADMIN) assign
  - `DELETE /api/v1/managers/{managerEmployeeId}/locations/{locationId}` (HR_ADMIN) unassign
- **Auth/security hardening implemented in this branch**
  - First-login password reset enforcement via `users.must_reset_password`
  - Liquibase migration: `003-add-users-must-reset-password.xml`
  - Endpoint: `POST /api/v1/auth/change-password-first-login`
  - Login blocked until reset when `must_reset_password=true`
  - Seeded HR admin is created with reset-required flag
  - Password complexity aligned across register and first-login reset
- **Code quality / review follow-ups**
  - Extracted duplicated principal parsing into shared `AuthenticationHelper`
  - Added missing 400 validation tests for location/department create endpoints

## Why
- Epic 3 requires location and department management plus manager location visibility:
  - `US-LOC-01`, `US-LOC-02`, `US-LOC-03`
  - `FR-LOC-01`, `FR-LOC-02`, `FR-LOC-03`
- `FR-LOC-03` states managers are assigned to locations, but assignment workflow was not explicitly defined in stories.  
  To make manager scoping usable end-to-end (and unblock downstream scheduling flows), HR assign/unassign endpoints were added.
- First-login reset flow was added to secure HR-provisioned credentials and seeded admin credentials.

## How to test
1. **Locations**
   - Login as HR_ADMIN
   - `POST /api/v1/locations` with `{name,address,maxHeadcountPerShift}` -> `201`
   - Repeat with same name -> `409`
   - `GET /api/v1/locations` -> `200`, active locations only
   - `PATCH /api/v1/locations/{id}` -> `200`
2. **Departments**
   - As HR_ADMIN: `POST /api/v1/locations/{locationId}/departments` with `{name}` -> `201`
   - Duplicate name in same location -> `409`
   - Invalid location -> `404`
   - As MANAGER: `GET /api/v1/locations/{locationId}/departments` -> `200`
3. **Manager assignments**
   - As HR_ADMIN: `POST /api/v1/managers/{managerEmployeeId}/locations/{locationId}` -> `201`
   - Duplicate assign -> `409`
   - As HR_ADMIN: `DELETE /api/v1/managers/{managerEmployeeId}/locations/{locationId}` -> `204`
   - Missing assignment on delete -> `404`
   - As MANAGER: `GET /api/v1/managers/me/locations` -> `200` with assigned list or `[]`
4. **First-login reset**
   - Create user via HR register or use seeded admin
   - Attempt login before reset -> blocked (reset required)
   - `POST /api/v1/auth/change-password-first-login` with valid current/new password -> `200`
   - Login with new password -> `200` + JWT
5. **Validation checks**
   - Create location with blank `name` -> `400` with `fieldErrors.name`
   - Create department with blank `name` -> `400` with `fieldErrors.name`
   - Weak new password on first-login reset -> `400` with `fieldErrors.newPassword`

## Notes
- **Gap addressed:** `FR-LOC-03` required manager-location assignment conceptually, but no explicit assignment story/endpoint was provided.  
  Added HR assign/unassign endpoints to make Epic 3 usable in practice and to avoid manual DB provisioning.
- **Approved scope deviation:** first-login password reset flow (`must_reset_password` + reset endpoint) was added though not explicitly listed in current stories, due to security requirements for HR-provisioned credentials and seeded admin credentials.
- **API versioning deviation maintained:** routes use `/api/v1/...` (not unversioned `/api/...`) to stay consistent with current project convention.

### CODE REVIEW

File: src/main/java/com/shiftsync/shiftsync/auth/service/impl/AuthServiceImpl.java [hunk 1]
```diff
@@ -1,6 +1,8 @@
 package com.shiftsync.shiftsync.auth.service.impl;
```
Comment (essilfie-b) [inline src/main/java/com/shiftsync/shiftsync/auth/service/impl/AuthServiceImpl.java:1]: **Bug: `mustResetPassword = true` set for all self-registering users**

Setting this flag in `register()` means every user who registers via `POST /api/v1/auth/register` will be blocked from logging in immediately after. This breaks US-AUTH-01 and US-AUTH-02.

The flag should only be set for HR-provisioned accounts (e.g. the `DataSeeder` seeded admin). Self-registered users chose their own password — they don't need a forced reset.

**Fix:** Remove `mustResetPassword(true)` from `register()`. The column default (`false`) and `@PrePersist` guard already handle the correct default.

**Review by essilfie-b — CHANGES_REQUESTED** [review]
## Overall: CHANGES REQUESTED

---

## What's good

- **PR description is excellent.** Has all four required sections (What, Why, How to test, Notes), references the correct FR/US numbers, and honestly flags the two scope deviations (assign/unassign endpoints and first-login reset). This is exactly the standard to maintain.
- **Reviewer requested correctly.** `@essilfie-b` is on the PR as required by the git workflow.
- **Branch name is correct.** `feature/manager-location-assignment` follows the `feature/` prefix convention.
- **Liquibase migration is well-written.** Column is `nullable=false` with `defaultValue=false`, and a rollback block is included. Clean.
- **Security config is correct.** `/api/v1/auth/change-password-first-login` is in `permitAll()` — users who cannot yet log in need to reach this endpoint unauthenticated.
- **Password complexity is consistent.** The same regex pattern is applied in both `RegisterRequest` and `ChangePasswordRequest`. Good.
- **Manager-location service validates roles.** Checking that the target employee actually has `MANAGER` role before assigning them to a location is the right guard.
- **Duplicate assignment returns 409, missing assignment on unassign returns 404.** Both are correct per the acceptance criteria pattern.
- **Swagger annotations on all new endpoints.** Every new endpoint has `@Operation` and `@ApiResponses`. Good discipline.

---

## Changes needed

### 1. Critical bug — `register()` sets `mustResetPassword = true` for ALL self-registering users

**File:** `src/main/java/com/shiftsync/shiftsync/auth/service/impl/AuthServiceImpl.java`

In the `register()` method, `mustResetPassword` is being set to `true` for every new user. This means anyone who calls `POST /api/v1/auth/register` — which is the standard self-registration flow — will be immediately blocked from logging in with "Password reset required before login".

This breaks **US-AUTH-01** and **US-AUTH-02** entirely. A user who just registered with their own chosen password has no reason to be forced through a reset flow.

The `mustResetPassword = true` flag should only be set in two places:
1. `DataSeeder.java` — for the seeded HR admin (already done correctly ✓)
2. Any future HR-provisioned account creation endpoint (not yet built)

**Fix:** Remove `mustResetPassword(true)` from `register()`. The default is already `false` via the `@PrePersist` guard and the Liquibase column default — rely on that.

---

### 2. No test files are visible in the changed files

The PR description states: *"Added missing 400 validation tests for location/department create endpoints"* under Code quality / review follow-ups.

No test files appear in the diff. If the tests were added, they need to be committed and included in this PR. If they were not added yet, remove that claim from the description to avoid misleading reviewers.

Per **NFR-05**, service layer coverage ≥ 80% is required, and at least three integration tests per major workflow. The Location/Department work in this PR is new and needs test coverage.

**Fix:** Either commit the tests and push them to this branch, or remove the claim from the PR description and open a follow-up `test/` branch immediately.

---

## Notes

You are clearly building momentum — the Epic 3 scope is well-reasoned, and proactively adding the assign/unassign endpoints to make `FR-LOC-03` usable end-to-end shows good product thinking. The transparency in the PR description about what was added beyond the stories is exactly the right habit.

The critical fix (item 1) is a one-line change — don't let it block you long. Address it, add the missing tests, and re-request review. You're close.

**Review by essilfie-b — APPROVED** [review]
## Overall: APPROVED

Solid PR. Epic 3 is fully delivered, the first-login reset flow is a sensible security hardening, and the PR description is the best one you've written so far — the What/Why/How to test/Notes format is exactly what the workflow requires. A few minor points below that are worth fixing forward but don't block merge.

---

## What's good

- PR description is exemplary — clear sections, FR/US references, explicit gap justification for the assign/unassign endpoints. Keep writing them this way.
- `feature/manager-location-assignment` branch name follows the workflow standard.
- All three user stories delivered: US-LOC-01, US-LOC-02, US-LOC-03.
- You correctly identified that FR-LOC-03 implies an assignment mechanism but no explicit story was written for it, and you addressed the gap with a clear justification in Notes. Good engineering judgement.
- `AuthenticationHelper` extraction removes the duplicated principal-parsing code that was scattered across controllers. This is the right refactor.
- `LocationServiceImpl.assignManagerToLocation` validates that the target employee actually has the MANAGER role before saving the assignment — important guard.
- Duplicate-assignment returns 409, missing-assignment on delete returns 404 — correct HTTP semantics throughout.
- `SecurityConfig` correctly permits `/change-password-first-login` without a token (the user can't log in to get a token yet, by design).
- `@Transactional` applied correctly on all write methods.
- Swagger docs present on every new endpoint.
- Tests added for auth, service, and controller layers.

---

## Changes needed

**1. `mustResetPassword = true` is set for every registration — confirm this is intentional scope**
`AuthServiceImpl.java`, inside `register()`:
```java
.mustResetPassword(true)
```
This means every account created via `POST /api/v1/auth/register` (which is HR_ADMIN only) will be blocked at login until they call `change-password-first-login`. That's fine for provisioning employees with temporary passwords, but be aware that if you ever want to let a user log in immediately after creation, you'd need to pass `mustResetPassword` as a parameter rather than hardcoding `true`. Not blocking — just make sure this matches the agreed behaviour.

**2. Redundant null-guard in `User.java` `@PrePersist`**
```java
@Builder.Default
private Boolean mustResetPassword = false;

@PrePersist
protected void onCreate() {
    if (mustResetPassword == null) {   // this can never be null — @Builder.Default covers it
        mustResetPassword = false;
    }
    ...
}
```
`@Builder.Default` already initialises the field. The null-guard in `@PrePersist` is dead code. Remove it to keep the class clean.

---

## Notes

- The `change-password-first-login` endpoint is unauthenticated by design — this is acceptable because it requires knowing the current password AND the account must already have `mustResetPassword=true`. There is no information leakage (all error paths return the same "Invalid credentials" message). If this pattern is reused elsewhere, document the reasoning with a brief comment so future maintainers understand the intent.
- The tests cover the happy path and key validation paths for both auth and location service changes. As you move into later weeks, make sure edge cases (e.g. assigning the same manager to a location twice, then deleting and reassigning) are also covered in integration tests.

Good work overall — Epic 3 is clean and ready.
